deploy/distributor.py

```diff
@@ -47,88 +54,103 @@ def __init__(self, web3, auction, auction_tx, auction_abi, distributor,
 
         # Start event watching
         self.watch_auction_bids()
-        self.watch_auction_end()
-        self.watch_auction_claim()
-        self.watch_auction_distributed()
+        # self.watch_auction_end()
+        # self.watch_auction_claim()
+        # self.watch_auction_distributed()
 
     def watch_auction_bids(self):
         self.filter_bids = self.handle_auction_logs('BidSubmission', self.add_address)
-        self.filter_bids.init()
+        self.filter_bids.init(self.watch_auction_end)
+        self.filter_bids.watch()
 
     def watch_auction_end(self):
         def set_end(event):
             self.auction_ended = True
+            self.final_price = self.auction.call().final_price()
 
         self.filter_auction_end = self.handle_auction_logs('AuctionEnded', set_end)
-        self.filter_auction_end.init()
+        self.filter_auction_end.init(self.watch_auction_claim)
+        self.filter_auction_end.watch()
 
     def watch_auction_claim(self):
-        # watch_logs(self.auction, 'ClaimedTokens', self.add_verified)
-
         self.filter_claims = self.handle_auction_logs('ClaimedTokens', self.add_verified)
-        self.filter_claims.init()
+        self.filter_claims.init(self.watch_auction_distributed)
+        self.filter_claims.watch()
 
     def watch_auction_distributed(self):
         def set_distribution_end(event):
             self.distribution_ended = True
             self.filter_bids.stop()
             self.filter_auction_end.stop()
-            print('set_distribution_end')
-
-        # watch_logs(self.auction, 'TokensDistributed', set_distribution_end)
-        print_logs(self.auction, 'TokensDistributed', 'DutchAuction')
-#        print_logs(self.auction, 'TradingStarted', 'DutchAuction')
 
         self.filter_distributed = self.handle_auction_logs('TokensDistributed',
                                                            set_distribution_end)
         self.filter_distributed.init()
+        self.filter_distributed.watch()
 
     def add_address(self, event):
         address = event['args']['_sender']
 
         # We might have multiple bids from the same bidder
-        if address not in self.addresses_claimable:
-            self.addresses_claimable.append(address)
-            self.values.append(0)
-            index = len(self.addresses_claimable) - 1
+        if address not in self.bidder_addresses:
```

Comment:
this would be much better done using dict mapping (a single dict where key is the address), to avoid computing of the indices and other array magic and to improve readability. If it works just leave it as it is now..